### PR TITLE
Added TxBody for advanced Era. WIP

### PR DESCRIPTION
**Pull request by TimSheard**
"Added the TxBody type with validity intervals and forge fields. Tied together with the Timelocks scripts. TxBody is newtype wrapped around a MemoBytes.
It exports a set of HasField instances appropriate for a TxBody. Also created a
test file with a minimum number of test, testing the HasField use and round-trip
CBOR properties. Pay close attention to the discussion about Mutually recursive
data types where this is a type family in the recursion cycle."

### CODE REVIEW

File: semantics/executable-spec/src/Data/MemoBytes.hs [hunk 1]
```diff
@@ -66,6 +75,9 @@ instance Show t => Show (MemoBytes t) where show (Memo y _) = show y
 
 instance Ord t => Ord (MemoBytes t) where compare (Memo x _) (Memo y _) = compare x y
 
+instance HasField tag t c => HasField (tag::Symbol) (MemoBytes t) c where
+   getField (Memo x _) = getField @tag x
```
Comment (nc6) [inline semantics/executable-spec/src/Data/MemoBytes.hs]: This is really nice :-)

File: shelley-ma/impl/cardano-ledger-shelley-ma.cabal [hunk 2]
```diff
@@ -24,6 +24,8 @@ library
     Cardano.Ledger.Mary
     Cardano.Ledger.ShelleyMA.Value
     Cardano.Ledger.ShelleyMA.Timelocks
+    Cardano.Ledger.ShelleyMA.TxBody
+    Cardano.Ledger.ShelleyMA.TestTxBody
```
Comment (nc6) [inline shelley-ma/impl/cardano-ledger-shelley-ma.cabal]: This should definitely move to a test package before merge

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs [hunk 3]
```diff
@@ -173,10 +195,8 @@ ininterval slot (SJust i_s, SJust i_f) = i_s <= slot && slot <= i_f
 -- will provide this operation. We will also need to correctly compute the witness
 -- set for this new (TxBody era) as well.
 
-type ValidityInterval = (StrictMaybe SlotNo, StrictMaybe SlotNo)
-
-txvld :: Core.TxBody era -> ValidityInterval
-txvld _ = (SNothing, SNothing)
+vldt :: Core.TxBody era -> ValidityInterval
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs]: I guess this can now be defined (with appropriate `HasField` constraint) - or maybe not necessary and we can just use the field directly?

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+import Shelley.Spec.Ledger.Keys (Hash)
+import Shelley.Spec.Ledger.MetaData (MetaDataHash)
+import Shelley.Spec.Ledger.PParams (Update)
+import Shelley.Spec.Ledger.Serialization (encodeFoldable)
+import Shelley.Spec.Ledger.TxBody (DCert (..), Wdrl (..))
+
+-- =========================================================================================================
+-- What do we require of TxBody? That it lives on the Block Chain, so it must have these instances.
+
+type ChainData t = (Eq t, Ord t, Show t, FromCBOR t, ToCBOR t, NoThunks t, NFData t, Typeable t)
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: Let's move this to `Cardano.Ledger.Core` or `Cardano.Ledger.Shelley` and use it for `TxBody` and `Value` as well

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+--type Value era = AbstractValue.Value era
+
+-- ================================================
+-- We cannot use the TxId from Shelley.Spec.Ledger.TxBody, because TxId and TxBody
+-- are mutually recursive (through the second (phantom) type index of Hash).
+-- TxId -> TxBody -> TxIn -> TxId. If we use the
+-- TxId from Shelley.Spec.Ledger.TxBody we get unsatisfiable constraints that
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: Jared also introduced `EraIndependentBlockBody` to get around this phantom type issue for the body hash, so an alternative would be to do the same here.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  fromCBOR = decode (RecD TxIn <! From <! From)
+
+-- ================================================
+-- We Avoid using TxOut from Shelley.Spec.Ledger.TxBody, because this involves convoluted
+-- use of the Compactible class, which only confuses things. Note that we don't know what
+-- (Value era) is, so we really can't say much about it. In order to choose a compactible
+-- instance for it if we have to know what (Value era) is, and we won't know that until a
+-- specific Era is chosen. But the code is supposed to abstract over every Era. Feel free
+-- to change this, but the intended use here is for (Value era) is to be
+-- (Cardano.Ledger.ShelleyMA.Value.Value era), which doesn't have a good compactible form,
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: At the moment we do rely on `Value` being compactible in other places, so this may be necessary anyway. And the same arguments that led to a compact form for `Coin` may call for a compact form for `Value` (we can at least squash the Ada in the same way)

File: semantics/executable-spec/src/Data/MemoBytes.hs (hunk 1 above)
Comment (TimSheard) [reply semantics/executable-spec/src/Data/MemoBytes.hs]: Works nicely in MemoBytes, not so nicely elsewhere.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs (hunk 3 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs]: This has been done by using a HasField constraint.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs (hunk 4 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: No problem. In the need I didn't need it, by reusing TxIn, TxId, and TxOut.
I had to do a bit of refactoring in module Shelley.Spec.Ledger.TxBody
to deal with the recursive cycle, but every thing worked out nicely.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs (hunk 6 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: By reusing TxId, TxIn and TxOut, all this goes away.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs (hunk 5 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: I introduced a nice abstraction parameterisng TxId by the phantom type.
Every thing becomes much simpler. This is well documented in the code now.

**Comment by TimSheard** [comment]
There is a recursive cycle in the types used to define TxBody
TxBody -> TxIn -> TxId -> TxBody

The invariant that these types enforce is that the hashed scripts in a TxId is a hash of the same type TxBody that the TxId occurs in. It would not be good if a hash of a Goguen body appeared in a TxId inside a Shelley TxBody.

We deal with this by Parameterize TxId by the type it should hash, and Parameterize TxIn with the same type which it passes to TxId's 

The approach (in a simplified form) looks like this

data TxId body era = TxId (Hash (Crypto era) body)
data TxIn body era = TxIn (TxId body era) Coin
data TxBody era = TxBody { inputs :: Set (TxIn (TxBody era) era), ...  }
                                                                  
Note how the phantom type (TxBody era) is passed through TxIn and then through TxId to become the phantom type of the embedded Hash.
Of course there are other complications due to efficiency, space use, and backward compatibility in the implementation in this branch but it really simplifies the approach to implementing TxBody for many Eras.

File: semantics/executable-spec/src/Data/MemoBytes.hs [hunk 7]
```diff
@@ -37,18 +40,24 @@ import Cardano.Binary
...
 
 -- ========================================================================
 
-data MemoBytes t = Memo {memotype :: !t, memobytes :: {-# UNPACK #-} !ShortByteString}
-  deriving (NoThunks) via AllowThunksIn '["memobytes"] (MemoBytes t)
+data MemoBytes t = Memo !t {-# UNPACK #-} !ShortByteString
+
+instance NoThunks t => NoThunks(MemoBytes t) where
+  noThunks ctxt (Memo x _) = noThunks ("Memo":ctxt) x
+  -- We deliberately allow thunks in the bytes
```
Comment (mrBliss) [inline semantics/executable-spec/src/Data/MemoBytes.hs]: What's the reason for this change? It was so pretty :cry: 

Moreover this manual instance is *wrong*. The [documentation of `NoThunks`](https://hackage.haskell.org/package/nothunks-0.1.1.0/docs/NoThunks-Class.html#v:noThunks) says:

> The default implementation of `noThunks` checks that the argument is in WHNF, and if so, adds the type into the context (using `showTypeOf`), and calls `wNoThunks`. See `ThunkInfo` for a detailed discussion of the type context.

So the method you should be overriding is `wNoThunks`, not `noThunks`. This is exactly why we provide the deriving-via wrappers like `AllowThunksIn` and `OnlyCheckWhnf`.

File: semantics/executable-spec/src/Data/MemoBytes.hs [hunk 8]
```diff
@@ -37,18 +40,24 @@ import Cardano.Binary
...
 import Prelude hiding (span)
+import GHC.TypeLits(Symbol)
+import GHC.Records
+import Codec.CBOR.Read(DeserialiseFailure,deserialiseFromBytes)
 
 -- ========================================================================
 
-data MemoBytes t = Memo {memotype :: !t, memobytes :: {-# UNPACK #-} !ShortByteString}
-  deriving (NoThunks) via AllowThunksIn '["memobytes"] (MemoBytes t)
+data MemoBytes t = Memo !t {-# UNPACK #-} !ShortByteString
```
Comment (mrBliss) [inline semantics/executable-spec/src/Data/MemoBytes.hs]: Wait, why the bang on the `ShortByteString` if the goal is to allow thunks in the bytes? The bang and the `UNPACK` pragma directly contradict the goal.

File: shelley-ma/impl/cardano-ledger-shelley-ma.cabal [hunk 9]
```diff
@@ -41,6 +44,9 @@ library
     nothunks,
     partial-order,
     shelley-spec-ledger,
-    small-steps
+    small-steps,
+    tasty-quickcheck,
+    tasty,
+    tasty-hunit
```
Comment (mrBliss) [inline shelley-ma/impl/cardano-ledger-shelley-ma.cabal]: Please don't add tests to the library

File: shelley-ma/impl/cardano-ledger-shelley-ma.cabal [hunk 10]
```diff
@@ -24,6 +24,8 @@ library
     Cardano.Ledger.Mary
     Cardano.Ledger.ShelleyMA.Value
     Cardano.Ledger.ShelleyMA.Timelocks
+    Cardano.Ledger.ShelleyMA.TxBody
+    -- Cardano.Ledger.ShelleyMA.TestTxBody
```
Comment (mrBliss) [inline shelley-ma/impl/cardano-ledger-shelley-ma.cabal]: I suppose this line only exists because this PR is WIP.

**Review by mrBliss — DISMISSED** [review]

File: semantics/executable-spec/src/Data/MemoBytes.hs (hunk 7 above)
Comment (TimSheard) [reply semantics/executable-spec/src/Data/MemoBytes.hs]: Thanks for pointing this out, the original need for this was some HasRecord instances did not let MemoBytes have its own record fields. But since we no longer need this, I have rolled it back. Also thanks reference to the video about NoThunks. Now I know why what we did there was bad.

File: shelley-ma/impl/cardano-ledger-shelley-ma.cabal (hunk 10 above)
Comment (TimSheard) [reply shelley-ma/impl/cardano-ledger-shelley-ma.cabal]: Yes, I set up a new test suite inside cardano-ledger/shelley-ma/impl and moved all the tests ther.

File: shelley-ma/impl/cardano-ledger-shelley-ma.cabal (hunk 9 above)
Comment (TimSheard) [reply shelley-ma/impl/cardano-ledger-shelley-ma.cabal]: I set up a separate test suite. so these are all gone.

File: semantics/executable-spec/src/Data/MemoBytes.hs (hunk 8 above)
Comment (TimSheard) [reply semantics/executable-spec/src/Data/MemoBytes.hs]: At one point we needed the MemoBytes type to not have any fileds of its own. So we changed this. That requirement is no longer needed, we also fixed the contradiction between the ! and the allowing thunks there, and reverted to the old style. Thanks for the advice and the link to the video.

File: semantics/executable-spec/src/Data/MemoBytes.hs (hunk 7 above)
Comment (mrBliss) [reply semantics/executable-spec/src/Data/MemoBytes.hs]: Great, looks better now :+1:

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+
+module Cardano.Ledger.ShelleyMA.TxBody
+  ( MaryBody (MaryBody, STxBody),
+    MaryBody' (..),
+  )
+where
+
+import Cardano.Binary (Annotator, FromCBOR (..), ToCBOR (..))
+-- \/ This is the type family Value
+import Cardano.Ledger.Core (CompactForm (..), Compactible (..), Value)
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: Use qualified import

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+-- ================================================
+-- There is a mutualy recursive cycle through the types
+-- TxId -> TxBody -> TxIn -> TxId.
+-- So we handle this by using the parameterized type
+-- (TxInParam (MaryBody era) era).  What this means that the
+-- TxOut's have TxId's can only have hashes of (MaryBody era)
+
+-- ===========================
+
+data MaryBody' era = MaryBody'
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: Why did you rename this? It was much better when it was just `TxBody`. The era is implied by the module it's in

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Value.hs [hunk 13]
```diff
@@ -67,7 +68,7 @@ newtype PolicyID era = PolicyID {policyID :: ScriptHash era}
 
 -- | The Value representing MultiAssets
 data Value era = Value !Integer !(Map (PolicyID era) (Map AssetID Integer))
-  deriving (Show, Generic)
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Value.hs]: What?! Why are we ordering `Value`s?

**Review by nc6 — CHANGES_REQUESTED** [review]
I think given the amount of change this import cycle is causing, we'd be better to be consistent and go with `EraIndependentTxHash` along the lines of the `EraIndependentBodyHash` which Jared introduced.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs (hunk 11 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: No need to use qualified import, as there are no conflicts.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Value.hs (hunk 13 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Value.hs]: Just a typo, It is fixed now.

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs (hunk 12 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/TxBody.hs]: Changed it back.

**Comment by TimSheard** [comment]
So I resolved the mutual recursion problem by using a technique pioneered by Jared. Introduce new uninhabited type for every instance of HashAnnotated.

data EraIndependentTx
data EraIndependentTxBody
data EraIndependentBlockBody
data EraIndependentWitVKey

The HashAnnotated class now has an associated type family HashIndex, and the method hashAnnotated is typed like this:

hashAnnotated :: a -> Hash.Hash (HASH (Crypto e)) (HashIndex a)

I cleaned up all of Jared's ad-hoc changes to be consistent with this paradigm. Including replacing eraIndependentTxHash  with  hashAnnotated.

This caused a few changes, but everything is cleaner, and the TxBody for the Mary era is now much cleaner as well. This should carry over to other Era's as well.

**Review by nc6 — APPROVED** [review]
Looks great! There's a couple of comment blocks with dead code that we should remove, but otherwise this looks good. I'll start rebasing atop this
